### PR TITLE
Addiing ByteDance Seed Seed-OSS

### PR DESCRIPTION
**Pull request by Fazziekey**
# What does this PR do?

This PR adds the support of codes for the coming Seed Oss models

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
@fazziekey looks exciting! Let us know when you need a review

**Comment by Fazziekey** [comment]
> @Fazziekey looks exciting! Let us know when you need a review

@Rocketknight1 Thanks! I finished most of the code and left the model IntegrationTest. You can review the model code

File: src/transformers/models/seed_oss/modular_seed_oss.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "ByteDance-Seed/SeedOss-36B"
+
+
+class SeedOssRMSNorm(LlamaRMSNorm):
+    pass
+
+
+class SeedOssMLP(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/seed_oss/modular_seed_oss.py:52]:
```suggestion
Starcoder2MLP
```
also ahs the residual dropout so I would just ensure our naming is the same! (your weight names are different so no need to change that)

File: src/transformers/models/seed_oss/modular_seed_oss.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        self.act_fn = ACT2FN[config.hidden_act]
+        self.dropout = nn.Dropout(config.resid_pdrop)
+
+    def forward(self, x):
+        down_proj = self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))
+        down_proj = self.dropout(down_proj)
+        return down_proj
+
+
+class SeedOssAttention(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/seed_oss/modular_seed_oss.py:70]: same here, `Starcoder2Attention` let's uniformize config names and good to go!

**Review by ArthurZucker — APPROVED** [review]
Damn ! I am bluffed almost perfect! 🤗

File: src/transformers/models/seed_oss/modular_seed_oss.py (hunk 1 above)
Comment (Fazziekey) [reply src/transformers/models/seed_oss/modular_seed_oss.py:52]: I find starcoder code 
```python3
class Starcoder2MLP(nn.Module):
    def __init__(self, config: Starcoder2Config):
        super().__init__()
        embed_dim = config.hidden_size
        self.c_fc = nn.Linear(embed_dim, config.intermediate_size, bias=config.use_bias)
        self.c_proj = nn.Linear(config.intermediate_size, embed_dim, bias=config.use_bias)
        self.act = ACT2FN[config.hidden_act]
        self.residual_dropout = config.residual_dropout

    def forward(self, hidden_states: Optional[tuple[torch.FloatTensor]]) -> torch.FloatTensor:
        hidden_states = self.c_fc(hidden_states)
        hidden_states = self.act(hidden_states)
        hidden_states = self.c_proj(hidden_states)
        hidden_states = nn.functional.dropout(hidden_states, p=self.residual_dropout, training=self.training)
        return hidden_states    
``` 

you mean use `nn.functional.dropout(hidden_states, p=self.residual_dropout, training=self.training) rather than self.dropout = nn.Dropout(config.resid_pdrop)?

Comment (Fazziekey) [reply src/transformers/models/seed_oss/modular_seed_oss.py:52]: actruely I follow the gpt2 code

Comment (Fazziekey) [reply src/transformers/models/seed_oss/modular_seed_oss.py:52]:
```python
class GPT2MLP(nn.Module):
    def __init__(self, intermediate_size, config):
        super().__init__()
        embed_dim = config.hidden_size
        self.c_fc = Conv1D(intermediate_size, embed_dim)
        self.c_proj = Conv1D(embed_dim, intermediate_size)
        self.act = ACT2FN[config.activation_function]
        self.dropout = nn.Dropout(config.resid_pdrop)

    def forward(self, hidden_states: Optional[tuple[torch.FloatTensor]]) -> torch.FloatTensor:
        hidden_states = self.c_fc(hidden_states)
        hidden_states = self.act(hidden_states)
        hidden_states = self.c_proj(hidden_states)
        hidden_states = self.dropout(hidden_states)
        return hidden_states
```

File: src/transformers/models/seed_oss/modular_seed_oss.py (hunk 2 above)
Comment (Fazziekey) [reply src/transformers/models/seed_oss/modular_seed_oss.py:70]: @ArthurZucker update to starCoder format

File: src/transformers/models/seed_oss/modular_seed_oss.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/seed_oss/modular_seed_oss.py:52]: sorry my suggestion was not absolutely necessary!

File: src/transformers/models/seed_oss/modular_seed_oss.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/seed_oss/modular_seed_oss.py:70]: thanks

File: src/transformers/models/seed_oss/modular_seed_oss.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/seed_oss/modular_seed_oss.py:52]: you can keep as is don't worry!

File: tests/models/seed_oss/test_modeling_seed_oss.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+@require_torch_large_accelerator
+class SeedOssIntegrationTest(unittest.TestCase):
+    input_text = ["Who Are you?", "Hi ByteDance-Seed"]
+    model_id = "ByteDance-Seed/SeedOss-36B"
+
+    def tearDown(self):
+        cleanup(torch_device, gc_collect=True)
+
+    def test_model_36b_fp16(self):
+        EXPECTED_TEXTS = ""
```
Comment (ArthurZucker) [inline tests/models/seed_oss/test_modeling_seed_oss.py]: let's update this and we can merge!

**Review by ArthurZucker — APPROVED** [review]

File: tests/models/seed_oss/test_modeling_seed_oss.py (hunk 3 above)
Comment (Fazziekey) [reply tests/models/seed_oss/test_modeling_seed_oss.py]: Hello, I add the test

Comment (Fazziekey) [reply tests/models/seed_oss/test_modeling_seed_oss.py]: fail at FAILED tests/models/blenderbot/test_modeling_blenderbot.py::BlenderbotStandaloneDecoderModelTest
image

File: src/transformers/models/seed_oss/modular_seed_oss.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        cache_position: Optional[torch.LongTensor] = None,
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        bsz, q_len, _ = hidden_states.size()
+
+        query_states = self.q_proj(hidden_states)
+        key_states = self.k_proj(hidden_states)
+        value_states = self.v_proj(hidden_states)
+
+        query_states = query_states.view(bsz, q_len, self.num_attention_heads, self.head_dim).transpose(1, 2)
```
Comment (LiuLi1998) [inline src/transformers/models/seed_oss/modular_seed_oss.py]: there should be query_states.view(bsz, q_len, -1, self.head_dim).transpose(1, 2), when I use vllm for inference, I have encounter some error cause by this line. vllm will cut num_attention_heads to num_attention_heads // tp_size

Comment (ArthurZucker) [reply src/transformers/models/seed_oss/modular_seed_oss.py]: you mean with transformers backend? You can use `-1` instead if it helps!

**Comment by ArthurZucker** [comment]
Happy to merge if ready for you! (the tests that don't have expected text need to be removed and good to go!

File: src/transformers/models/seed_oss/modular_seed_oss.py (hunk 4 above)
Comment (vasqu) [reply src/transformers/models/seed_oss/modular_seed_oss.py]: actually quite important, our own TP stack might fail here as well then :eyes: 

Meaning all `.view`s here need to take `(bsz, q_len, -1, self.head_dim)` instead

Comment (Fazziekey) [reply src/transformers/models/seed_oss/modular_seed_oss.py]:
```
        input_shape = hidden_states.shape[:-1]
        hidden_shape = (*input_shape, -1, self.head_dim)

        query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)
        key_states = self.k_proj(hidden_states).view(hidden_shape).transpose(1, 2)
        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)
```

Comment (Fazziekey) [reply src/transformers/models/seed_oss/modular_seed_oss.py]: updated

**Comment by Fazziekey** [comment]
> Happy to merge if ready for you! (the tests that don't have expected text need to be removed and good to go!

I think all is ready, can you merge it? @ArthurZucker  Thanks so much

**Comment by Fazziekey** [comment]
@ArthurZucker hello, all is ready, can you merge it?

**Comment by ArthurZucker** [comment]
Yes super sorry!

**Comment by ArthurZucker** [comment]
Merged!
